### PR TITLE
[MIG] stock_request: Migration to 16.0

### PR DESCRIPTION
**Pull request by nicomacr**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please include https://github.com/OCA/stock-logistics-warehouse/pull/1628, although the migration scripts are not needed this time.

**Comment by nicomacr** [comment]
> Please include [OCA/stock-logistics-warehouse#1628](https://github.com/OCA/stock-logistics-warehouse/pull/1628), although the migration scripts are not needed this time.

Done! @pedrobaeza

File: migrations/15.0.1.4.0/noupdate_changes.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Comment (pedrobaeza) [inline migrations/15.0.1.4.0/noupdate_changes.xml]: You haven't removed the migration scripts from this commit.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -101,7 +101,7 @@ repos:
       - id: pyupgrade
         args: ["--keep-percent-format"]
   - repo: https://github.com/PyCQA/isort
-    rev: 5.10.1
+    rev: 5.12.0
```
Comment (pedrobaeza) [inline .pre-commit-config.yaml]: Why changing this manually? Let me instead update the copier template

Comment (pedrobaeza) [reply .pre-commit-config.yaml]: #2 Rebase when I merge it.

Comment (pedrobaeza) [reply .pre-commit-config.yaml]: Done

**Comment by pedrobaeza** [comment]
I observe a problem in the history fetching that will affect later blaming. I think the problem comes from using the command:
```
git format-patch --keep-subject --stdout origin/16.0..origin/15.0 -- $module | git am -3
```
instead of:
```
git format-patch --keep-subject --stdout origin/16.0..origin/15.0 -- $module | git am -3 --keep --directory $module
```

I think it should be tackled.

**Comment by nicomacr** [comment]
> I observe a problem in the history fetching that will affect later blaming. I think the problem comes from using the command:
> 
> ```
> git format-patch --keep-subject --stdout origin/16.0..origin/15.0 -- $module | git am -3
> ```
> 
> instead of:
> 
> ```
> git format-patch --keep-subject --stdout origin/16.0..origin/15.0 -- $module | git am -3 --keep --directory $module
> ```
> 
> I think it should be tackled.

Understood, what actions do you suggest to do? do again the migration of the module using that command?

**Comment by pedrobaeza** [comment]
You can:

- Save current module folder to any place.
- Start from scratch doing the procedure.
- Remove the module folder and copy the stored one.
- Commit such changes.
- Forced push to this branch.

**Comment by nicomacr** [comment]
@pedrobaeza Ready the changes

**Comment by pedrobaeza** [comment]
Ouch, now you face the contrary one: some contents were put inside a double folder. I don't understand why it wasn't correct then nor now...

File: stock_request/models/stock_request.py [hunk 3]
```diff
@@ -21,12 +34,7 @@ def _get_expected_date():
 
     name = fields.Char(states={"draft": [("readonly", False)]})
     state = fields.Selection(
-        selection=[
-            ("draft", "Draft"),
-            ("open", "In progress"),
-            ("done", "Done"),
-            ("cancel", "Cancelled"),
-        ],
+        selection=_get_request_states,
```
Comment (rousseldenis) [inline stock_request/models/stock_request.py]: Prefer a lambda here to help override and put fileds definition on top

File: stock_request/models/stock_request.py [hunk 4]
```diff
@@ -5,13 +5,26 @@
 from odoo.exceptions import UserError, ValidationError
 from odoo.tools import float_compare
 
+REQUEST_STATES = [
+    ("draft", "Draft"),
```
Comment (rousseldenis) [inline stock_request/models/stock_request.py]: If you don't use selection attribute directly, you should use the translate function.

But I see this approach has been finally removed : https://github.com/OCA/stock-logistics-warehouse/pull/1622/files

So, please get the previous code back

File: stock_request/models/stock_request_abstract.py [hunk 5]
```diff
@@ -238,23 +231,33 @@ def onchange_location_id(self):
                 self.warehouse_id = loc_wh
                 self.with_context(no_change_childs=True).onchange_warehouse_id()
 
+    @api.onchange("allow_virtual_location")
+    def onchange_allow_virtual_location(self):
+        if self.allow_virtual_location:
+            return {"domain": {"location_id": []}}
```
Comment (rousseldenis) [inline stock_request/models/stock_request_abstract.py]: I'm not in favor reintroducing domains in onchanges as this is deprecated in core and will be removed in later version.

File: stock_request/models/stock_request_abstract.py [hunk 6]
```diff
@@ -238,23 +231,33 @@ def onchange_location_id(self):
...
-                    "|",
-                    ("company_id", "=", False),
-                    ("company_id", "=", self.company_id.id),
-                ],
-                limit=1,
+                [("company_id", "=", self.company_id.id)], limit=1
             )
             self.onchange_warehouse_id()
 
+        return {"domain": {"warehouse_id": [("company_id", "=", self.company_id.id)]}}
```
Comment (rousseldenis) [inline stock_request/models/stock_request_abstract.py]: Same as above

**Review by rousseldenis — CHANGES_REQUESTED** [review]
I would say domains in onchanges is the major thing to not reintroduce

File: stock_request/models/stock_request.py (hunk 3 above)
Comment (pedrobaeza) [reply stock_request/models/stock_request.py]: Please don't. No need of that to insert elements where you want. See https://github.com/OCA/stock-logistics-warehouse/blob/4e338d152d79d99e298ccf34f548515a3a7aa8f7/stock_request_submit/models/stock_request.py#L11

and if you use a method, the translations are very bad handled.

File: stock_request/models/stock_request.py (hunk 4 above)
Comment (pedrobaeza) [reply stock_request/models/stock_request.py]: The same stated above.

File: stock_request/models/stock_request_abstract.py (hunk 5 above)
Comment (pedrobaeza) [reply stock_request/models/stock_request_abstract.py]: I agree here. This should be handed on a static domain, but you can use simple expressions on domains. Example: https://github.com/OCA/rma/blob/e0c17a60e152f8814242e026166e1f2114789944/rma_sale/models/rma.py#L41-L42

File: stock_request/models/stock_request.py (hunk 4 above)
Comment (rousseldenis) [reply stock_request/models/stock_request.py]: @nicomacr Please changes this according to what pointed above:

image

File: stock_request/models/stock_request.py [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+            raise ValidationError(_("Company must be equal to the order"))
+
+    @api.constrains("order_id", "expected_date")
+    def check_order_expected_date(self):
+        if self.order_id and self.order_id.expected_date != self.expected_date:
+            raise ValidationError(_("Expected date must be equal to the order"))
+
+    @api.constrains("order_id", "picking_policy")
+    def check_order_picking_policy(self):
+        if self.order_id and self.order_id.picking_policy != self.picking_policy:
```
Comment (rousseldenis) [inline stock_request/models/stock_request.py]: @nicomacr For every constraint, change to multi recordset

**Comment by rousseldenis** [comment]
@nicomacr I still observe a problem with the commit history about the directory structure.

Could you please redo migration procedure from scratch ? Thanks

**Comment by rousseldenis** [comment]
@nicomacr

**Comment by EvaSForgeFlow** [comment]
Hi @nicomacr, I think when you access to Settings you are not able to see the name of the module:

error_on_config

To avoid this, in the **res_config_settings_views.xml** file include in here:

`<div
                    class="app_settings_block"
                    data-string="Stock Request"
                    data-key="stock_request"
                    groups="stock_request.group_stock_request_manager"
                >`

the module string, as follows:

`<div class="app_settings_block" data-string="Stock Request" string="Stock Request" data-key="stock_request" groups="stock_request.group_stock_request_manager">`

image

**Comment by nicomacr** [comment]
@rousseldenis @pedrobaeza sorry for the delay. I've just updated this PR with a new code from scratch.

**Comment by JordiMForgeFlow** [comment]
Hi @nicomacr 

Could you rebase the PR? Thank you! :)

**Comment by rousseldenis** [comment]
@nicomacr The correct command to move a module from a repo to another should be:

We assert 'origin' is your branch repository based on stock-logistics-request.

Add the OCA repository in your configuration:

`git remote add OCA https://github.com/OCA/stock-logistics-request`
`git fetch OCA`

Then, here is the trick. Add the former repository remote:

`git remote add OCA-warehouse https://github.com/OCA/stock-logistics-warehouse`
`git fetch OCA-warehouse`

Then do the 'normal' migration command, adapting the source remote:

` git format-patch --keep-subject --stdout OCA/16.0..OCA-warehouse/15.0 -- stock_request | git am -3 --keep`


And voilà! No need to create a temporary directory

**Comment by rousseldenis** [comment]
I assert that before doing those commands, you have a clean branch reset on OCA/16.0

**Comment by nicomacr** [comment]
@rousseldenis Yes I've made this using that command. I've just fixed the problems now. It's all green

**Comment by rousseldenis** [comment]
Mmmmh I don't think so as I see weird commit :

image

**Comment by rousseldenis** [comment]
And in migration commit, you have several files locations changes :

image

**Comment by nicomacr** [comment]
@rousseldenis Yes, You're right about it, I've had a dirty repo. Now I did it again. and I think it's OK. Please tell me if you see that is OK or not.

**Comment by matt454357** [comment]
Hi @rousseldenis,
I believe the requested changes have been made. Can this PR be merged?

**Review by rousseldenis — APPROVED** [review]
Code review. @nicomacr Next migration, put pre-commit stuff in a separate commit.

**Comment by rousseldenis** [comment]
@JordiBForgeFlow @pedrobaeza Maybe you want to review this ?

File: stock_request/models/stock_location_route.py [hunk 8]
```diff
@@ -6,7 +6,7 @@
 
 
 class StockLocationRoute(models.Model):
-    _inherit = "stock.location.route"
+    _inherit = "stock.route"
```
Comment (victoralmau) [inline stock_request/models/stock_location_route.py]: Rename the class and file name

File: stock_request/tests/test_stock_request.py [hunk 9]
```diff
@@ -1,38 +1,33 @@
 # Copyright 2017 ForgeFlow S.L.
-# Copyright 2022-2023 Tecnativa - Víctor Martínez
 # License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl-3.0).
 
 from collections import Counter
 from datetime import datetime
 
 from odoo import exceptions, fields
-from odoo.tests import common, new_test_user
+from odoo.tests import common
```
Comment (victoralmau) [inline stock_request/tests/test_stock_request.py:8]: Do not delete `new_test_user` (it is quite useful and you can define as many fields as you want) to create `_create_user()`

File: stock_request/tests/test_stock_request.py [hunk 10]
```diff
@@ -147,23 +182,10 @@ def _create_product_attribute_value(self, name, attribute):
...
-    def _create_location_route(self, **vals):
-        return self.env["stock.location.route"].create(
-            dict(
-                product_categ_selectable=False,
-                product_selectable=True,
-                sequence=10,
-                **vals
-            )
-        )
-
```
Comment (victoralmau) [inline stock_request/tests/test_stock_request.py]: Is it necessary to eliminate these methods?

File: stock_request/tests/test_stock_request.py [hunk 11]
```diff
@@ -0,0 +1,1231 @@
...
+            {"name": name, "attribute_id": attribute}
+        )
+
+    def _create_product_attribute(self, name):
+        return self.env["product.attribute"].create({"name": name})
+
+
+class TestStockRequestBase(TestStockRequest):
+    def setUp(self):
+        super(TestStockRequestBase, self).setUp()
```
Comment (victoralmau) [inline stock_request/tests/test_stock_request.py]:
```suggestion
        super().setUp()
```

File: stock_request/tests/test_stock_request.py [hunk 12]
```diff
@@ -0,0 +1,1231 @@
...
+
+from collections import Counter
+from datetime import datetime
+
+from odoo import exceptions, fields
+from odoo.tests import common
+
+
+class TestStockRequest(common.TransactionCase):
+    def setUp(self):
```
Comment (victoralmau) [inline stock_request/tests/test_stock_request.py]:
```suggestion
    @classmethod
    def setUpClass(cls):
```

**Review by victoralmau — CHANGES_REQUESTED** [review]
Please remove migration scripts from 15.0

I have checked how it works in runboat, but it is not possible to create a stock request if the user does not have units of measure permissions (the `product_uom_id` field is mandatory but not self-defined, something that in v15 worked correctly). can you check it?

**Comment by pedrobaeza** [comment]
> I have checked how it works in runboat, but it is not possible to create a stock request if the user does not have units of measure permissions (the `product_uom_id` field is mandatory but not self-defined, something that in v15 worked correctly). can you check it?

That's because the migration guide hasn't been followed, specifically https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0/_compare/ad52290cdb5f5385ae25fad0c4862a4fa244d9f8...e2d0ba6cd1bfdc8cee787c3fd61b6feecfbe49cc

File: stock_request/views/stock_request_order_views.xml [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+                            >
+                                <tree editable="bottom">
+                                    <field name="name" readonly="1" />
+                                    <field name="product_id" />
+                                    <field name="allowed_uom_categ_id" invisible="1" />
+                                    <field
+                                        name="product_uom_id"
+                                        options="{'no_open': True, 'no_create': True}"
+                                        groups="uom.group_uom"
+                                    />
```
Comment (victoralmau) [inline stock_request/views/stock_request_order_views.xml:134]:
```suggestion
                                    <field
                                        name="product_uom_id"
                                        options="{'no_open': True, 'no_create': True}"
                                        groups="uom.group_uom"
                                    />
                                    <field name="product_uom_id" readonly="True" groups="!uom.group_uom" force_save="1"/>
```

File: stock_request/views/stock_request_views.xml [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+                            <field name="allowed_uom_categ_id" invisible="1" />
+                            <label for="product_uom_qty" />
+                            <div>
+                                <field name="product_uom_qty" class="oe_inline" />
+                                <field
+                                    name="product_uom_id"
+                                    class="oe_inline"
+                                    options="{'no_open': True, 'no_create': True}"
+                                    groups="uom.group_uom"
+                                />
```
Comment (victoralmau) [inline stock_request/views/stock_request_views.xml:214]:
```suggestion
                                <field
                                    name="product_uom_id"
                                    class="oe_inline"
                                    options="{'no_open': True, 'no_create': True}"
                                    groups="uom.group_uom"
                                />
                                <field name="product_uom_id" class="oe_inline" readonly="True" groups="!uom.group_uom" force_save="1"/>
```

**Review by victoralmau — CHANGES_REQUESTED** [review]
Please remove migration scripts from 15.0 and fix test errors.

File: stock_request/tests/test_stock_request.py [hunk 15]
```diff
@@ -0,0 +1,1237 @@
...
+                        "product_uom_qty": 1.0,
+                        "company_id": cls.main_company.id,
+                        "warehouse_id": cls.warehouse.id,
+                        "location_id": cls.warehouse.lot_stock_id.id,
+                        "expected_date": expected_date,
+                    },
+                ),
+            ],
+        }
+        cls.order = cls.request_order.new(vals)
```
Comment (victoralmau) [inline stock_request/tests/test_stock_request.py]:
```suggestion
        cls.order = cls.request_order.create(vals)
```

**Review by victoralmau — CHANGES_REQUESTED** [review]

File: stock_request/tests/test_stock_request.py (hunk 15 above)
Comment (rousseldenis) [reply stock_request/tests/test_stock_request.py]: @victoralmau Sometimes, create is not necessary, especially in tests to improve performances.

**Comment by rousseldenis** [comment]
@nicomacr Could you check tests ?

**Comment by victoralmau** [comment]
Please, cherry-pick https://github.com/OCA/stock-logistics-warehouse/pull/1885 to the commit history

**Comment by gcaceres93** [comment]
Hello, is there any further advances on this pull request?

**Comment by victoralmau** [comment]
Superseed by https://github.com/OCA/stock-logistics-request/pull/13

**Comment by rousseldenis** [comment]
Closing as no answer for a long time

**Pull request closed without merging**